discovery-provider/src/queries/attestation.py

```diff
@@ -0,0 +1,155 @@
+
+# Define custom error strings
+CHALLENGE_INCOMPLETE = "CHALLENGE_INCOMPLETE"
+ALREADY_DISBURSED = "ALREADY_DISBURSED"
+INVALID_ORACLE = "INVALID_ORACLE"
+MISSING_CHALLENGES = "MISSING_CHALLENGES"
+INVALID_INPUT = "INVALID_INPUT"
+
+
+class AttestationError(Exception):
+    pass
+
+
+def is_valid_oracle(address: str):
+    # TODO: flesh this out, refresh oracle addresses [AUD-729]
+    default_oracle_address = shared_config["discprov"]["default_oracle_address"]
+    return address == default_oracle_address
+
+
+def sign_attestation(attestation_str: str, private_key: str):
+    to_sign_hash = Web3.keccak(text=attestation_str).hex()
+    encoded_to_sign = encode_defunct(hexstr=to_sign_hash)
+    signed_message = w3.eth.account.sign_message(
+        encoded_to_sign, private_key=private_key
+    )
+    return signed_message.signature.hex()
+
+
+def get_attestation(
+    session: Session,
+    *,
+    challenge_id: str,
+    user_id: int,
+    oracle_address: str,
+    specifier: str,
+):
+    """
+    Returns a owner_wallet, signed_attestation tuple,
+    or throws an error explaining why the attestation was
+    not able to be created."""
+    if not user_id or not challenge_id or not oracle_address:
+        raise AttestationError(INVALID_INPUT)
+
+    # First, validate the oracle adddress
+    if not is_valid_oracle(oracle_address):
+        raise AttestationError(INVALID_ORACLE)
+
+    challenges_and_disbursements: List[
+        Tuple[UserChallenge, Challenge, ChallengeDisbursement]
+    ] = (
+        session.query(UserChallenge, Challenge, ChallengeDisbursement)
+        .join(Challenge, Challenge.id == UserChallenge.challenge_id)
+        # Need to do outerjoin because some challenges
+        # may not have disbursements
+        .outerjoin(
+            ChallengeDisbursement,
+            and_(
+                ChallengeDisbursement.specifier == UserChallenge.specifier,
+                ChallengeDisbursement.challenge_id == UserChallenge.challenge_id,
+            ),
+        )
+        .filter(
+            UserChallenge.user_id == user_id,
+            UserChallenge.challenge_id == challenge_id,
+            UserChallenge.specifier == specifier,
+        )
+    ).all()
+
+    if not len(challenges_and_disbursements) == 1:
+        raise AttestationError(MISSING_CHALLENGES)
+    user_challenge, challenge, disbursement = (
+        challenges_and_disbursements[0][0],
+        challenges_and_disbursements[0][1],
+        challenges_and_disbursements[0][2],
+    )
+    if not user_challenge.is_complete:
+        raise AttestationError(CHALLENGE_INCOMPLETE)
+    if disbursement:
+        raise AttestationError(ALREADY_DISBURSED)
+
+    # Get the users's eth address
+    user_eth_address = (
+        session.query(User.wallet)
+        .filter(
+            User.is_current == True,
+            User.user_id == user_id,
+        )
+        .one_or_none()
+    )
+    if not user_eth_address:
+        raise AttestationError(MISSING_CHALLENGES)
```

Comment:
MISSING_ETH_ADDRESS?

Reply:
this is such a weird edge case that should never happen (was just being extra thorough), don't think it's worth creating it's own error